### PR TITLE
backend/pref: added aadhaar hash

### PR DESCRIPTION
**Pull request by AshishSaini25**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [x] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

**Review by piyushKumar-1 — APPROVED** [review]
What if a user actually wants to register new account using same aadhar number ? Maybe lets create a unlinking dashboard api or implement a ui flow to unlink ?

File: Backend/app/dashboard/CommonAPIs/src/Dashboard/ProviderPlatform/Driver.hs [hunk 1]
```diff
@@ -185,11 +185,20 @@ data DriverAadhaarInfoRes = DriverAadhaarInfoRes
...
   }
   deriving stock (Eq, Show, Generic)
   deriving anyclass (ToJSON, FromJSON, ToSchema)
 
+-- driver aadhaar Info api by mobile Number ----------------------------------------
+
+type DriverAadhaarInfoByPhoneAPI =
+  Capture "mobileNo" Text
+    :> "aadhaarInfobyMobileNumber"
+    :> Get '[JSON] DriverAadhaarInfoRes
```
Comment (hkmangla) [inline Backend/app/dashboard/CommonAPIs/src/Dashboard/ProviderPlatform/Driver.hs:198]: Just create new variable named `type DriverAadhaarInfoByPhoneReq = DriverAadhaarInfoRes ` and use that.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Driver.hs [hunk 2]
```diff
@@ -698,3 +700,40 @@ clearOnRideStuckDrivers merchantShortId = do
...
+
+---------------------------------------------------------------------
+
+---------------------------------------------------------------------
+
+updateDriverAadhaar :: ShortId DM.Merchant -> Text -> Common.UpdateDriverDataReq -> Flow APISuccess
+updateDriverAadhaar merchantShortId phoneNumber req = do
+  mobileNumberHash <- getDbHash phoneNumber
+  aadhaarNumberHash <- getDbHash req.driverAadhaarNumber
+  aadhaarInfo <- AV.findByAadhaarNumberHash aadhaarNumberHash
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Driver.hs:731]: Aadhaar info will always be there, right?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Driver.hs [hunk 3]
```diff
@@ -698,3 +700,40 @@ clearOnRideStuckDrivers merchantShortId = do
...
+  when (isJust aadhaarInfo) $ throwError AadhaarAlreadyLinked
+  merchant <- findMerchantByShortId merchantShortId
+  driver <- QPerson.findByMobileNumberAndMerchant "+91" mobileNumberHash merchant.id >>= fromMaybeM (InvalidRequest "Person not found")
+  Esq.runTransaction $ AV.findByPhoneNumberAndUpdate req.driverName req.driverGender req.driverDob (Just aadhaarNumberHash) req.isVerified driver.id
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Driver.hs:735]: rename function to `updateByPhoneNumber`

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Driver.hs (hunk 2 above)
Comment (hkmangla) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Driver.hs:731]: As we are using the same table.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/DriverOnboarding/AadhaarVerification.hs [hunk 4]
```diff
@@ -56,6 +65,9 @@ generateAadhaarOtp isDashboard mbMerchant personId req = do
   driverInfo <- DriverInfo.findById (cast personId) >>= fromMaybeM (PersonNotFound personId.getId)
   when driverInfo.blocked $ throwError DriverAccountBlocked
   when (driverInfo.aadhaarVerified) $ throwError AadhaarAlreadyVerified
+  aadhaarHash <- getDbHash req.aadhaarNumber
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/DriverOnboarding/AadhaarVerification.hs:68]: Just create one function checkForDuplicacy and use that.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/DriverOnboarding/AadhaarVerification.hs [hunk 5]
```diff
@@ -40,8 +42,33 @@ findByDriverId driverId = do
...
     return aadhaar
 
+findByAadhaarNumberHash ::
+  Transactionable m =>
+  DbHash ->
+  m (Maybe AadhaarVerification)
+findByAadhaarNumberHash aadhaarHash = do
+  findOne $ do
+    aadhaar <- from $ table @AadhaarVerificationT
+    where_ $ aadhaar ^. AadhaarVerificationAadhaarNumberHash ==. val (Just aadhaarHash)
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/DriverOnboarding/AadhaarVerification.hs:52]: Create a DB migration for adding index on aadhaar hash.

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Driver.hs (hunk 2 above)
Comment (AshishSaini25) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Driver.hs:731]: yes but let say that aadhaar is already linked with some other account then we  should not allow to update that

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/DriverOnboarding/AadhaarVerification.hs [hunk 6]
```diff
@@ -53,9 +62,10 @@ generateAadhaarOtp ::
   Flow AadhaarVerification.AadhaarVerificationResp
 generateAadhaarOtp isDashboard mbMerchant personId req = do
   person <- Person.findById personId >>= fromMaybeM (PersonNotFound personId.getId)
-  driverInfo <- DriverInfo.findById (cast personId) >>= fromMaybeM (PersonNotFound personId.getId)
-  when driverInfo.blocked $ throwError DriverAccountBlocked
-  when (driverInfo.aadhaarVerified) $ throwError AadhaarAlreadyVerified
+  checkForDuplicacy person.id
+  aadhaarHash <- getDbHash req.aadhaarNumber
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/DriverOnboarding/AadhaarVerification.hs:66]: You moved the wrong code inside checkFor Duplicacy.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/DriverOnboarding/c061ac91-6cc9-4bf4-821b-064f5a01a5dd.hs [hunk 7]
```diff
@@ -0,0 +1,224 @@
+{-
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/DriverOnboarding/c061ac91-6cc9-4bf4-821b-064f5a01a5dd.hs:1]: What is this?

**Review by hkmangla — CHANGES_REQUESTED** [review]

**Review by hkmangla — APPROVED** [review]

**Pull request merged**